Separate federation build.sh into development and deployment scripts.

madhusudancs opened the pull request:
The idea behind this separation is that it provides a clear distinction
between the dev environment and the prod environment. The
deploy/deploy.sh script will be shipped to the users, but
develop/develop.sh will be purely for development purposes and won't
be part of a release distribution.

Purely for developer convenience, all the deployment functionality is
made available through the develop/develop.sh script.

This change also copies deploy/\* files into the release distribution.

cc @kubernetes/sig-cluster-federation @colhom 

``` release-note
Federation can now be deployed using the `federation/deploy/deploy.sh` script. This script does not depend on any of the development environment shell library/scripts. This is an alternative to the current `federation-up.sh`/`federation-down.sh` scripts. Both the scripts are going to co-exist in this release, but the `federation-up.sh`/`federation-down.sh` scripts might be removed in a future release in favor of `federation/deploy/deploy.sh` script.
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30744)

<!-- Reviewable:end -->

madhusudancs committed:
Separate federation build.sh into development and deployment scripts.

The idea behind this separation is that it provides a clear distinction
between the dev environment and the prod environment. The
deploy/deploy.sh script will be shipped to the users, but
develop/develop.sh will be purely for development purposes and won't
be part of a release distribution.

Purely for developer convenience, all the deployment functionality is
made available through the develop/develop.sh script.

This change also copies deploy/* files into the release distribution.

nikhiljindal commented on federation/develop/develop.sh:
```diff
@@ -0,0 +1,218 @@
+
+readonly TMP_DIR="$(mktemp -d)"
+readonly FEDERATION_OUTPUT_ROOT="${LOCAL_OUTPUT_ROOT}/federation"
+readonly VERSIONS_FILE="${FEDERATION_OUTPUT_ROOT}/versions"
```
To make sure I understand, this file is doing what federated-image.tag does currently, right?

nikhiljindal commented on federation/develop/develop.sh:
```diff
@@ -0,0 +1,218 @@
+# KUBE_REGISTRY=localhost:5000/anushku \
+# KUBE_VERSION=1.3.0-dev ./build.sh build_image
+#
+# will build the Docker images with the specified repository name and the
```
build and push

nikhiljindal replied:
and actually deploy as well. ok, we cant mention everything here.
Fine with whatever you think is reasonable.

nikhiljindal commented on federation/deploy/deploy.sh:
```diff
@@ -0,0 +1,141 @@
+  "init")
+  pull_installer
+  ;;
+  "deploy_clusters")
```
What do you think about adding a "deploy" action that deploys all kube clusters and federation control plane. Same for "destroy"?
Fine to add in a follow up PR.

nikhiljindal commented on federation/deploy/deploy.sh:
```diff
@@ -0,0 +1,141 @@
+
+function ensure_files() {
+  kube::log::status "Ensure provider is supported"
+  if [[ "${KUBERNETES_PROVIDER:-}" != "gce" ]]; then
```
I thought GKE was supported as well. Is it not?

nikhiljindal commented on build/common.sh:
```diff
@@ -1029,10 +1029,11 @@ function kube::release::package_full_tarball() {
   mkdir "${release_stage}/federation"
   cp -R "${KUBE_ROOT}/federation/cluster" "${release_stage}/federation/"
   cp -R "${KUBE_ROOT}/federation/manifests" "${release_stage}/federation/"
+  cp -R "${KUBE_ROOT}/federation/deploy" "${release_stage}/federation"
```
follow the slash at end convention?

nikhiljindal commented:
Thanks, mostly looks good. A few minor comments, mostly asking for clarifications.

madhusudancs committed:
Addressed review comments.

madhusudancs commented:
@nikhiljindal addressed your comments. PTAL.

---

Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[build/common.sh, line 1036 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIb-r1-1036:-KQ2CYDq67GPx_cvhSNz:b-dvhy26) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/build/common.sh#L1036)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> follow the slash at end convention?
</details>
Thanks for pointing that out. That slash is important.

---

*[federation/deploy/deploy.sh, line 55 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIg-r1-55:-KQ2D3YFl5SApv0U4tnJ:bsjq8hr) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/deploy/deploy.sh#L55)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> I thought GKE was supported as well. Is it not?
</details>
Sure. Added. Thanks.

---

*[federation/deploy/deploy.sh, line 124 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIg-r1-124:-KQ2D7LtEtJkoFUs310F:b-r08n1r) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/deploy/deploy.sh#L124)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> What do you think about adding a "deploy" action that deploys all kube clusters and federation control plane. Same for "destroy"?
> 
> Fine to add in a follow up PR.
</details>
I would like them to be separate here, so that people always think about them as two separate phases. It is easy to wrap them together, just two commands, instead of one. Our development Makefile does this.

Also, I am considering the possibility of allowing multiple commands in one invocation. For example, users can run 

```shell
$ ./deploy.sh init deploy_clusters deploy_federation
```
in one go.

---

*[federation/develop/develop.sh, line 24 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIh-r1-24:-KQ2DrN5QygjcoKWpxtv:by5zfmu) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/develop/develop.sh#L24)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> and actually deploy as well. ok, we cant mention everything here.
> 
> Fine with whatever you think is reasonable.
</details>
The example command just builds the images. It neither pushes nor deploys.

---

*[federation/develop/develop.sh, line 44 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIh-r1-44:-KQ2DyKrfB3fGqpuFq6d:bvwokr2) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/develop/develop.sh#L44)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> To make sure I understand, this file is doing what federated-image.tag does currently, right?
</details>
Correct, but also in a more machine-readable fashion.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-:-KQ2EA83iD-gV8nc5wws:b-6rd8t9)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[federation/develop/develop.sh, line 44 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIh-r1-44:-KQ2rRia5ecqcaMcQaMp:bvbq6t9) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/develop/develop.sh#L44)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> Correct, but also in a more machine-readable fashion.
</details>
Why is that? Isnt the content same? Just store the version?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[federation/develop/develop.sh, line 44 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIh-r1-44:-KQ2sUbL2YC7MezmdC3N:bf05h45) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/develop/develop.sh#L44)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> Why is that? Isnt the content same? Just store the version?
</details>
It is just the version right now. I have plans to add other values. For example, one of them is the registry/image name.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[federation/deploy/deploy.sh, line 124 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIg-r1-124:-KQ2sg6hPhIta8lw5sg7:b-7siri) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/deploy/deploy.sh#L124)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> I would like them to be separate here, so that people always think about them as two separate phases. It is easy to wrap them together, just two commands, instead of one. Our development Makefile does this.
> 
> Also, I am considering the possibility of allowing multiple commands in one invocation. For example, users can run 
> 
> ```shell
> $ ./deploy.sh init deploy_clusters deploy_federation
> ```
> in one go.
</details>
The readme still says "deploy" will deploy both kubernetes and federation clusters

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[federation/deploy/deploy.sh, line 124 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIg-r1-124:-KQ2tKi4dm0J4GLiHYnv:b-n0snih) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/deploy/deploy.sh#L124)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> The readme still says "deploy" will deploy both kubernetes and federation clusters
</details>
README wasn't updated in this PR. If you insist, I can send those changes in this PR.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744)*
<!-- Sent from Reviewable.io -->

madhusudancs commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[federation/deploy/deploy.sh, line 124 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIg-r1-124:-KQ2tab1cyObPwi8Wko2:bfbn9zh) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/deploy/deploy.sh#L124)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> README wasn't updated in this PR. If you insist, I can send those changes in this PR.
</details>
Btw, I forgot to mention that the README in the federation directory is not a user facing doc, it is for developers. And since developers are likely going to drive this through the make command and since we haven't made any changes to the Make command in this PR, it does not make that much sense to update the README here.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[federation/deploy/deploy.sh, line 124 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIg-r1-124:-KQJzZuSdlRJ0ZiWrfCH:b-m0ui1c) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/deploy/deploy.sh#L124)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> Btw, I forgot to mention that the README in the federation directory is not a user facing doc, it is for developers. And since developers are likely going to drive this through the make command and since we haven't made any changes to the Make command in this PR, it does not make that much sense to update the README here.
</details>
I pointed it out because before these changes "deploy" action worked as written in the README.
This change removes the support for deploy, so it seems correct to me to update the README as well.
(Tough what I will really prefer is to update the code to continue supporting the "deploy" action but I am fine since it seems you disagree).

---

*[federation/develop/develop.sh, line 44 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIh-r1-44:-KQJxhRo4pjfszJxRcY0:bdlqxjn) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/develop/develop.sh#L44)):*
<details><summary><i>Previously, madhusudancs (Madhusudan.C.S) wrote…</i></summary>
> It is just the version right now. I have plans to add other values. For example, one of them is the registry/image name.
</details>
Sorry, still didnt understand why is this more machine-readable than federated-image.tag :) (but its a side discussion, not blocking)

Also, is `versions` the right name if you already have plans to store more info than just the image version? Or you plan to rename the file later?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744)*
<!-- Sent from Reviewable.io -->

madhusudancs committed:
Update README.

madhusudancs commented:
@nikhiljindal Addressed the comments. PTAL.

---

Review status: 0 of 7 files reviewed at latest revision, 5 unresolved discussions.

---

*[federation/deploy/deploy.sh, line 124 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIg-r1-124:-KQMQ3_TqtVS78YLaybD:b-4eh3na) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/deploy/deploy.sh#L124)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> I pointed it out because before these changes "deploy" action worked as written in the README.
> This change removes the support for deploy, so it seems correct to me to update the README as well.
> (Tough what I will really prefer is to update the code to continue supporting the "deploy" action but I am fine since it seems you disagree).
> 
</details>
It makes zero-sense to update the README here. But I have updated it here, since you insisted. PTAL.

I split these changes into multiple PRs to make the life of reviewers somewhat easier so that they could review small independent changes separately. So this is part of an ongoing work, this is not the final PR in the work. 

---

*[federation/develop/develop.sh, line 44 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-KPM7Ecq5CVSUAedCiIh-r1-44:-KQMRCArhn2lSaxnyTs5:besufii) ([raw file](https://github.com/kubernetes/kubernetes/blob/e52484ff378af6f9ebb216e9fe242b4b683214e5/federation/develop/develop.sh#L44)):*
> Sorry, still didnt understand why is this more machine-readable than federated-image.tag :) (but its a side discussion, not blocking)

Because it is JSON. There are tools built to read and/or modify it.

> Also, is versions the right name if you already have plans to store more info than just the image version? Or you plan to rename the file later?

Naming is hard :) Let's figure out the appropriate name, when that time comes around.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30744#-:-KQMRiNHc4actXnNLOdX:bcu4awj)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented on federation/README.md:
```diff
@@ -17,23 +17,42 @@ as explained in that guide, you also need to install [`jq`](https://stedolan.git
+
+To initialize the deployment run:
+
+(This pull the installer images)
```
pulls

nikhiljindal commented:
LGTM, thanks

nikhiljindal commented on federation/deploy/deploy.sh:
```diff
@@ -0,0 +1,141 @@
+# Provides logging facilities
+source "${KUBE_ROOT}/cluster/lib/logging.sh"
+
+readonly KUBE_ANYWHERE_FEDERATION_IMAGE="gcr.io/madhusudancs-containers/kubernetes-anywhere-federation"
```
Can you add a comment regarding where does the code for this image exist?
Also, why cant we use the vanila kubernetes-anywhere image?
